File: 044_Two Sum.py
"""
Two Sum

Find indices of two numbers that add up to a specific target.
"""

from typing import List

def two_sum(nums: List[int], target: int) -> List[int]:
    m = {}
    for i, v in enumerate(nums):
        need = target - v
        if need in m:
            return [m[need], i]
        m[v] = i
    return []


if __name__ == '__main__':
    print('TwoSum 1:', two_sum([2,7,11,15], 9))
    print('TwoSum 2:', two_sum([3,2,4], 6))

# Time Complexity: O(n)
# Space Complexity: O(n)

File: 044_Two Sum.cpp
/*
Two Sum

Find indices of two numbers that add up to a specific target.
*/

#include <bits/stdc++.h>
using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int,int> m;
    for (int i = 0; i < (int)nums.size(); ++i) {
        int need = target - nums[i];
        if (m.count(need)) return {m[need], i};
        m[nums[i]] = i;
    }
    return {};
}

int main() {
    vector<int> a1 = {2,7,11,15};
    auto r1 = twoSum(a1, 9);
    cout << "TwoSum 1:";
    for (int x: r1) cout << ' ' << x;
    cout << endl; // 0 1

    vector<int> a2 = {3,2,4};
    auto r2 = twoSum(a2, 6);
    cout << "TwoSum 2:";
    for (int x: r2) cout << ' ' << x;
    cout << endl; // 1 2
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)
